libsolidity/analysis/OverflowChecker.cpp

```diff
@@ -0,0 +1,220 @@
+		m_varOverflowInfo[var] = bounds;
+		if (!bounds->fitsType(*(var->annotation().type)))
+			m_errorReporter.warning(_assignment.location(), "Variable assignment may cause overflow.  " + bounds->toString());
+		else
+			m_errorReporter.warning(_assignment.location(), "Variable assignment ok.  Range: "+bounds->toString()+" Storage: "+var->annotation().type->toString(false));
+	}
+
+	return false;
+}
+
+bool OverflowChecker::lockConstantExpression(Expression const& _expression)
+{
+	auto const* rationalNumberType = dynamic_cast<RationalNumberType const*>(_expression.annotation().type.get());
+	if (rationalNumberType)
+	{
+		shared_ptr<IntegerOverflowBounds> bounds = make_shared<IntegerOverflowBounds>(*rationalNumberType);
+		m_overflowInfo[&_expression] = bounds;
+		return true;
+	}
+	return false;
+}
+
+bool OverflowChecker::visit(Literal const& _literal)
+{
+	lockConstantExpression(_literal);
+	return true;
+}
+
+bool OverflowChecker::visit(BinaryOperation const& _operation)
+{
+	if (lockConstantExpression(_operation))
+		return false; // Do not descend
+	else
+	{
+		// Need to do more work...
+		auto const* integerType = dynamic_cast<IntegerType const*>(_operation.annotation().commonType.get());
+		if (integerType)
+		{
+			//First, descend and resolve both sides of the operation.
+			_operation.leftExpression().accept(*this);
+			_operation.rightExpression().accept(*this);
+
+			//If the operator is a comparison, then generate any possible constraints.
+			//The most naive form of this will only work if the LHS is an identifier.
+			if (Token::isCompareOp(_operation.getOperator()))
+			{
+				auto const identifier = dynamic_cast<Identifier const*>(&_operation.leftExpression());
+				if (identifier)
+				{
+					auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
+					if (var)
+					{
+						IntegerOverflowBounds constraint(*_operation.leftExpression().annotation().type);
+						constraint.setToMaximumRange(*_operation.leftExpression().annotation().type);
+						constraint.performBinaryOp(_operation.getOperator(),
+								constraint,
+								*(m_overflowInfo[&(_operation.rightExpression())]));
+						m_constraints[var] = make_shared<IntegerOverflowBounds>(constraint);
+						return true;
+					}
+				}
+			}
+
+
+			IntegerOverflowBounds bounds(*integerType);
+			shared_ptr<OverflowBounds> leftBounds = m_overflowInfo[&(_operation.leftExpression())];
+			shared_ptr<OverflowBounds> rightBounds = m_overflowInfo[&(_operation.rightExpression())];
+
+			bounds.performBinaryOp(_operation.getOperator(), *leftBounds, *rightBounds);
+			m_overflowInfo[&_operation] = make_shared<IntegerOverflowBounds>(bounds);
+			if (!bounds.fitsType(*integerType))
+				m_errorReporter.warning(_operation.location(), "Expression potentially overflowing its storage.  Range is: " + bounds.toString() + " Storage is: "+integerType->toString(false));
```

Comment:
Please don`t use the word storage here, it has a special meaning in Solidity :-)